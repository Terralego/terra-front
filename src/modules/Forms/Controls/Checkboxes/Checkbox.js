import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox as BPCheckbox } from '@blueprintjs/core';

const Checkbox = ({ checked, label, onToggle, value }) => (
  <BPCheckbox
    onChange={() => onToggle(value)}
    label={label || value}
    checked={checked}
  />
);

Checkbox.propTypes = {
  value: PropTypes.string,
  label: PropTypes.string,
  checked: PropTypes.bool,
  onToggle: PropTypes.func,
};

Checkbox.defaultProps = {
  value: '',
  label: '',
  checked: false,
  onToggle () {},
};

export default Checkbox;
