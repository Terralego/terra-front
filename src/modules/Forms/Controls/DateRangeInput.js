import React from 'react';
import classnames from 'classnames';

import PropTypes from 'prop-types';

import { Spinner } from '@blueprintjs/core';
import { DateRangeInput as BPDateRangeInput } from '@blueprintjs/datetime';

import localeUtils from '../../../utils/localeUtils';
import translateMock from '../../../utils/translate';

// Date Format
const formatDate = date => date.toLocaleDateString();
const parseDate = str => new Date(str);

export const DateRangeInput = ({
  className,
  label,
  value,
  onChange,
  translate,
  locale = global.navigator.language,
  shortcuts = false,
  loading,
  ...props
}) => (
  <div
    className="control-container"
  >
    <p className="control-label">{label}</p>
    {!!loading && <Spinner size={20} />}
    {!loading && (
      <BPDateRangeInput
        className={classnames('tf-control-date-range', className)}
        formatDate={formatDate}
        onChange={onChange}
        parseDate={parseDate}
        value={value}
        overlappingDatesMessage={translate('terralego.forms.controls.daterangeinput.overlapping_dates')}
        invalidDateMessage={translate('terralego.forms.controls.daterangeinput.invalid_date')}
        popoverProps={{ usePortal: false }}
        localeUtils={localeUtils}
        locale={locale.substr(0, 2)}
        shortcuts={shortcuts}
        startInputProps={{
          placeholder: translate('terralego.forms.controls.daterangeinput.start_placeholder'),
          className: 'tf-control-date-range__start-input',
        }}
        endInputProps={{
          placeholder: translate('terralego.forms.controls.daterangeinput.end_placeholder'),
          className: 'tf-control-date-range__en d-input',
        }}
        {...props}
      />
    )}
  </div>
);

DateRangeInput.propTypes = {
  label: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.arrayOf(PropTypes.instanceOf(Date)),
  translate: PropTypes.func,
  loading: PropTypes.bool,
};

DateRangeInput.defaultProps = {
  label: '',
  value: [null, null],
  onChange () {},
  translate: translateMock({
    'terralego.forms.controls.daterangeinput.overlapping_dates': 'Overlapping date',
    'terralego.forms.controls.daterangeinput.invalid_date': 'Invalid date',
    'terralego.forms.controls.daterangeinput.start_placeholder': 'Start date',
    'terralego.forms.controls.daterangeinput.end_placeholder': 'End date',
  }),
  loading: false,
};

export default DateRangeInput;
