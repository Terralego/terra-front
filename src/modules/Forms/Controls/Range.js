import PropTypes from 'prop-types';
import React from 'react';
import { Spinner } from '@blueprintjs/core';

import './index.scss';
import RangeNumeric from './RangeNumeric';
import RangeSlider from './RangeSlider';

export const DEFAULT_MIN = 0;
export const DEFAULT_MAX = 100;

/**
 * Range component that allows editing either by a slider or 2 numeric inputs
 */
export const Range = props => {
  const {
    label,
    min,
    max,
    threshold,
    stepSize,
    value,
    loading,
  } = props;
  const Component = (max - min) / stepSize > threshold ? RangeNumeric : RangeSlider;

  return (
    <div className="control-container control--range range">
      <p className="control-label">{label}</p>
      {loading ? <Spinner size={20} /> : (
        <Component
          {...props}
          value={(!Array.isArray(value) || value.length !== 2) ? [min, max] : value}
        />
      )}
    </div>
  );
};

Range.propTypes = {
  label: PropTypes.string,
  value: PropTypes.arrayOf(PropTypes.number),
  onChange: PropTypes.func,
  min: PropTypes.number,
  max: PropTypes.number,
  /** Number of steps for when to switch to 2 numeric inputs */
  threshold: PropTypes.number,
  stepSize: PropTypes.number,
  loading: PropTypes.bool,
};

Range.defaultProps = {
  label: '',
  value: undefined,
  onChange () {},
  min: DEFAULT_MIN,
  max: DEFAULT_MAX,
  threshold: 250,
  stepSize: 1,
  loading: false,
};

export default Range;
