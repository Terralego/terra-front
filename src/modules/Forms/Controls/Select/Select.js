import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { MenuItem, Button, Spinner } from '@blueprintjs/core';
import { Select as BPSelect } from '@blueprintjs/select';
import uuid from 'uuid/v4';

import { preventEnterKeyPress } from '../../../../utils/event';
import translateMock from '../../../../utils/translate';
import formatValues from '../formatValues';
import NoValues from '../NoValues';

const MAX_NON_FILTERABLE_COUNT = 9;

export class Select extends React.Component {
  static propTypes = {
    label: PropTypes.string,
    onChange: PropTypes.func,
    values: PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
          label: PropTypes.string.isRequired,
          value: PropTypes.any.isRequired,
        }),
      ]),
    ).isRequired,
    isSubmissionPrevented: PropTypes.bool,
    translate: PropTypes.func,
    loading: PropTypes.bool,
  };

  static defaultProps = {
    label: '',
    isSubmissionPrevented: true,
    onChange () {},
    translate: translateMock({
      'terralego.forms.controls.select.empty_item': 'All',
      'terralego.forms.controls.select.no_results': 'No results.',
      'terralego.forms.controls.select.select_placeholder': 'Enter a query first',
      'terralego.forms.controls.select.input_placeholder': 'Filter...',
      'terralego.forms.controls.noValues': 'No choice available',
    }),
    loading: false,
  };

  static getDerivedStateFromProps ({ values = [], value }) {
    if (!values.length) {
      return {
        values,
        value: null,
      };
    }

    const formatedValues = formatValues(values);

    const foundValue = value
      ? formatedValues.find(({ value: availableVal }) => value === availableVal)
      : formatedValues[0];

    return {
      values: formatedValues,
      value: foundValue,
    };
  }

  state = {
    values: [],
    value: null,
  };

  uuid = `select-${uuid()}`;

  handleChange = ({ value }) => {
    const { onChange } = this.props;
    onChange(value);
  };

  render () {
    const {
      label,
      placeholder,
      isSubmissionPrevented,
      value: origValue,
      className,
      inline,
      fullWidth,
      translate,
      loading,
      ...props
    } = this.props;

    const { values = [], value } = this.state;
    const { handleChange } = this;

    const displayedValue = value ? value.label : '';

    return (
      <div
        className={classnames(
          'control-container',
          'control-container--select',
          { 'control-container--full-width': fullWidth },
        )}
        onKeyPress={isSubmissionPrevented ? preventEnterKeyPress : null}
        role="presentation"
      >
        <label
          htmlFor={this.uuid}
          className={classnames(
            'control-label',
            { 'control-label--inline': inline },
          )}
        >
          {label}
        </label>
        {loading && <Spinner size={20} />}
        {!loading && !values.length && <NoValues placeholder={translate('terralego.forms.controls.noValues')} />}
        {!loading && (
          <BPSelect
            className={classnames('tf-select', className)}
            popoverProps={{ usePortal: false, boundary: 'viewport', minimal: true }}
            items={values}
            filterable={values.length > MAX_NON_FILTERABLE_COUNT}
            onQueryChange={this.handleQueryChange}
            initialContent={(values && values.length > 250)
              ? <MenuItem disabled text={translate('terralego.forms.controls.select.select_placeholder')} />
              : undefined}
            inputProps={{
              placeholder: translate('terralego.forms.controls.select.input_placeholder'),
            }}
            itemPredicate={
              (query, { label: itemLabel }) => itemLabel.toLowerCase().includes(query.toLowerCase())
            }
            itemRenderer={({
              label: itemLabel, value: itemValue,
            }, {
              handleClick, modifiers: { matchesPredicate, ...modifiers },
            }) => (
              <div
                key={`${itemValue}${itemLabel}`}
                className={classnames({
                  'control-container__item': true,
                  'control-container__item--empty': itemValue === '',
                })}
              >
                <MenuItem
                  {...modifiers}
                  active={value.value === itemValue}
                  onClick={handleClick}
                  text={itemLabel}
                />
              </div>
            )}
            noResults={(
              <MenuItem
                disabled
                text={translate('terralego.forms.controls.select.no_results')}
              />
            )}
            onItemSelect={handleChange}
            {...props}
          >
            <Button
              id={this.uuid}
              text={displayedValue || translate('terralego.forms.controls.select.empty_item')}
              rightIcon="double-caret-vertical"
            />
          </BPSelect>
        )}
      </div>
    );
  }
}

export default Select;
