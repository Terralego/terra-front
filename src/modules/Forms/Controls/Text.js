import React from 'react';
import PropTypes from 'prop-types';
import { InputGroup } from '@blueprintjs/core';

import './index.scss';

export const Text = ({ label, onChange, value }) => (
  <div className="control-container">
    <p className="control-label">{label}</p>
    <InputGroup
      onChange={({ target: { value: v } }) => onChange(v)}
      value={value}
    />
  </div>
);

Text.propTypes = {
  label: PropTypes.string,
  onChange: PropTypes.func,
  value: PropTypes.string,
};

Text.defaultProps = {
  label: '',
  onChange () {},
  value: '',
};

export default Text;
