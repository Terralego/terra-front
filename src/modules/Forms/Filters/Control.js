import React from 'react';
import moize from 'moize';
import PropTypes from 'prop-types';
import translateMock from '../../../utils/translate';

import { TYPE_MANY } from './Filters';

const getValuesWithEmptyItem = moize((values, emptySelectItem) => [{
  value: '',
  label: emptySelectItem,
}, ...values]);

const defaultTranslate = translateMock({
  'terralego.forms.controls.generic.empty_item': 'Nothing',
});

export class Control extends React.Component {
  static propTypes = {
    translate: PropTypes.func,
  };

  static defaultProps = {
    translate: defaultTranslate,
  };

  static getDerivedStateFromProps (props) {
    const {
      values,
      type,
      translate,
    } = props;
    if (!values) return null;

    const withEmptyValue = values && type !== TYPE_MANY;

    return {
      values: withEmptyValue
        ? getValuesWithEmptyItem(values, translate('terralego.forms.controls.generic.empty_item'))
        : values,
    };
  }

  state = {
    values: [],
  };

  render () {
    const { component: Component, translate, ...props } = this.props;
    const { values } = this.state;

    return (
      <Component
        translate={translate === defaultTranslate ? undefined : translate}
        {...props}
        values={values}
      />
    );
  }
}

export default Control;
