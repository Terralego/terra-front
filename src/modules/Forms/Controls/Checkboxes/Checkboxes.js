import React from 'react';
import PropTypes from 'prop-types';
import { Spinner } from '@blueprintjs/core';

import Checkbox from './Checkbox';
import formatValues from '../formatValues';
import NoValues from '../NoValues';
import '../index.scss';

export class Checkboxes extends React.Component {
  static getDerivedStateFromProps ({ values }) {
    if (!values) return null;

    return {
      values: formatValues(values),
    };
  }

  state = {}

  onToggle = toggledValue => {
    const { value, onChange } = this.props;
    const newValue = value.includes(toggledValue)
      ? [...value.filter(val => val !== toggledValue)]
      : [...value, toggledValue];
    onChange(newValue);
  }

  render () {
    const { value: mainValue, label: mainLabel, loading } = this.props;
    const { values } = this.state;
    const { onToggle } = this;

    return (
      <div className="control-container">
        <p className="control-label">{mainLabel}</p>
        {loading && <Spinner size={20} />}
        {!loading && !values.length && <NoValues />}
        {!loading && values.map(({ label, value }) => (
          <Checkbox
            key={`${label}${value}`}
            onToggle={onToggle}
            value={value}
            label={label}
            checked={mainValue.includes(value)}
          />
        ))}
      </div>
    );
  }
}

Checkboxes.propTypes = {
  label: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.string), PropTypes.string]),
  // Prop is used in getDerivedStateFromProps
  // eslint-disable-next-line react/no-unused-prop-types
  values: PropTypes.arrayOf(PropTypes.string),
  onChange: PropTypes.func,
  loading: PropTypes.bool,
};

Checkboxes.defaultProps = {
  label: '',
  value: [],
  values: [],
  onChange () {},
  loading: false,
};

export default Checkboxes;
