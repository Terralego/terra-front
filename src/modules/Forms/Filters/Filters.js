import React from 'react';
import PropTypes from 'prop-types';

import Select from '../Controls/Select';
import Text from '../Controls/Text';
import Checkboxes from '../Controls/Checkboxes';
import Range from '../Controls/Range';
import Switch from '../Controls/Switch';
import MultiSelect from '../Controls/MultiSelect';
import DateRangeInput from '../Controls/DateRangeInput';
import Control from './Control';
import translateMock from '../../../utils/translate';

export const TYPE_SINGLE = 'single';
export const TYPE_MANY = 'many';
export const TYPE_RANGE = 'range';
export const TYPE_BOOL = 'boolean';

export function getComponent (type, values, display, format) {
  switch (type) {
    case TYPE_SINGLE:
      return Array.isArray(values)
        ? Select
        : Text;
    case TYPE_MANY:
      return (
        Array.isArray(values) && values.length > 10) ||
          (display === 'select')
        ? MultiSelect
        : Checkboxes;
    case TYPE_RANGE:
      return format === 'date'
        ? DateRangeInput
        : Range;
    case TYPE_BOOL:
      return Switch;
    default:
      return null;
  }
}

export const Filters = ({ properties, filters, translate, onChange }) => (
  <div>
    {filters.map(({ type, values, property, display, format, label, ...props }) => (
      <Control
        {...props}
        label={label}
        translate={translate}
        component={getComponent(type, values, display, format)}
        key={property}
        type={type}
        values={values}
        property={property}
        onChange={value => onChange({ ...properties, [property]: value })}
        value={properties[property]}
      />
    ))}
  </div>
);

Filters.propTypes = {
  onChange: PropTypes.func,
  filters: PropTypes.arrayOf(PropTypes.shape({
    property: PropTypes.string.isRequired,
    label: PropTypes.string,
    placeholder: PropTypes.string,
    type: PropTypes.oneOf([
      TYPE_SINGLE,
      TYPE_MANY,
      TYPE_RANGE,
      TYPE_BOOL,
    ]).isRequired,
    values: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.string),
      PropTypes.arrayOf(PropTypes.number),
      PropTypes.bool,
      PropTypes.arrayOf(PropTypes.shape({
        label: PropTypes.string,
        value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
      })),
    ]),

  })),
  properties: PropTypes.shape({
    // property: value
  }),
  translate: PropTypes.func,
};

Filters.defaultProps = {
  onChange () {},
  filters: [],
  properties: {},
  translate: translateMock(),
};

export default Filters;
