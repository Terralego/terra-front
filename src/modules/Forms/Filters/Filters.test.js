import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';

import Select from '../Controls/Select';
import Text from '../Controls/Text';
import Checkboxes from '../Controls/Checkboxes';
import Range from '../Controls/Range';
import Switch from '../Controls/Switch';

import Filters, { TYPE_SINGLE, TYPE_MANY, TYPE_RANGE, TYPE_BOOL, getComponent } from './Filters';
import DateRangeInput from '../Controls/DateRangeInput';

jest.mock('@blueprintjs/core', () => ({
  Checkbox: () => <p>Checkbox</p>,
  InputGroup: () => <p>Input</p>,
  Select: ({ children }) => children,
  Switch: () => <p>Switch</p>,
  MenuItem: ({ text = 'Menu Item' }) => <p>{text}</p>,
  Button: () => <p>Button</p>,
  RangeSlider: () => <p>Range</p>,
}));

jest.mock('../Controls/Text', () => () => <p>ControlText</p>);
jest.mock('../Controls/MultiSelect', () => () => <p>ControlMultiSelect</p>);
jest.mock('../Controls/Select', () => ({ initialContent }) => (
  <p>
    {initialContent ? (
      <>
        ControlSelect with initialContent set to
        {initialContent}
      </>
    ) : 'ControlSelect'}
  </p>
));
jest.mock('../Controls/Checkboxes', () => () => <p>ControlCheckboxes</p>);
jest.mock('../Controls/Range', () => () => <p>ControlRange</p>);
jest.mock('../Controls/DateRangeInput', () => () => <p>DateRangeInput</p>);

it('should build a form', () => {
  const tree = renderer.create((
    <Filters
      onChange={() => null}
      filters={[{
        property: 'single_value',
        label: 'Vocations',
        type: TYPE_SINGLE,
      }, {
        property: 'single_value_forced',
        label: 'Vocations',
        type: TYPE_SINGLE,
        values: ['développeur', 'UX designer', 'Chef de projet'],
      }, {
        property: 'shitload_of_values',
        label: 'Shitload',
        type: TYPE_SINGLE,
        display: 'select',
        values: Array(251),
      }, {
        property: 'many_values',
        label: 'Foo',
        type: TYPE_MANY,
        values: ['développeur', 'UX designer', 'Chef de projet'],
      }, {
        property: 'many_values_select_forced',
        label: 'Foo',
        type: TYPE_MANY,
        display: 'select',
        values: ['développeur', 'UX designer', 'Chef de projet'],
      }, {
        property: 'range_values',
        label: 'Percentage of bar',
        type: TYPE_RANGE,
        values: [40, 60],
        min: 0,
        max: 100,
      }, {
        property: 'one_value',
        label: 'Booléen',
        type: TYPE_BOOL,
        value: true,
      }]}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should mount & update correctly', () => {
  const onChange = jest.fn();
  const wrapper = shallow((
    <Filters
      onChange={onChange}
      filters={[{
        property: 'single_value',
        label: 'Vocations',
        type: TYPE_SINGLE,
      }, {
        property: 'single_value_forced',
        label: 'Vocations',
        type: TYPE_SINGLE,
        values: ['développeur', 'UX designer', 'Chef de projet'],
      }, {
        property: 'many_values',
        label: 'Foo',
        type: TYPE_MANY,
        values: ['développeur', 'UX designer', 'Chef de projet'],
      }, {
        property: 'range_values',
        label: 'Percentage of bar',
        type: TYPE_RANGE,
        values: [40, 60],
        min: 0,
        max: 100,
      }, {
        property: 'switch_value',
        label: 'Switch label',
        type: TYPE_BOOL,
      }]}
    />
  ));

  wrapper.instance().onChange('pwout')('pwet');
  expect(onChange).toHaveBeenCalledWith({ pwout: 'pwet' });
});

it('should get component', () => {
  expect(getComponent(TYPE_SINGLE, 'text') === Text).toBe(true);
  expect(getComponent(TYPE_SINGLE, ['text', 'text2']) === Select).toBe(true);
  expect(getComponent(TYPE_MANY, ['text', 'text2']) === Checkboxes).toBe(true);
  expect(getComponent(TYPE_BOOL, true) === Switch).toBe(true);
  expect(getComponent(TYPE_RANGE, [10, 90]) === Range).toBe(true);
  expect(getComponent(TYPE_RANGE, [null, null], null, 'date') === DateRangeInput).toBe(true);
  expect(getComponent('NONE_EXISTING_TYPE', 'text') === null).toBe(true);
});

it('should ensure translation works', () => {
  const translate = jest.fn();
  const wrapper = shallow((
    <Filters
      translate={translate}
      filters={[{
        property: 'switch_value',
        label: 'Switch label',
        type: TYPE_BOOL,
      }]}
    />
  ));
  expect(wrapper.find('Control').props().translate).toEqual(translate);
});
