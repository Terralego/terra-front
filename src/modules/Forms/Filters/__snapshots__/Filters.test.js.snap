// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should build a form 1`] = `
<div>
  <p>
    ControlText
  </p>
  <p>
    ControlSelect
  </p>
  <p>
    ControlSelect
  </p>
  <p>
    ControlCheckboxes
  </p>
  <p>
    ControlMultiSelect
  </p>
  <p>
    ControlRange
  </p>
  <div
    className="control-container"
  >
    <p>
      Switch
    </p>
  </div>
</div>
`;

exports[`should render loading correctly 1`] = `
<div>
  <p>
    ControlSelect
  </p>
</div>
`;
